unistd/pwd.c

```diff
@@ -24,69 +25,108 @@
 
 struct passwd pwnam;
 char pw_name[NAME_MAX];
+char intstr[16];
 char pw_gecos[128];
 char pw_dir[PATH_MAX];
 char pw_shell[PATH_MAX];
 char pw_passwd[128];
 
 
-struct passwd *getpwnam(const char *name)
+static int readpwentry(FILE *fp, char *pwentry, unsigned int maxlen, unsigned int islast)
+{
+	int c = 0;
+	unsigned int i = 0, isbad = 0;
+
+	do {
+		c = fgetc(fp);
+		if (c == ':' || c == '\n' || c == EOF)
+			break;
+
+		if (i < maxlen - 1)
+			pwentry[i++] = c;
+		else
+			isbad = 1;
+	} while (!isbad);
+	pwentry[i] = '\0';
+
+	if ((c == EOF || c == '\n') && !islast)
+		isbad = 1;
+
+	return isbad;
+}
+
+
+static inline int readuid(int readpwentry, uid_t *uid)
 {
-	FILE *fp;
-	char buf[1024];
-
-	if ((fp = fopen(PASSWD_PATH, "r")) == NULL)
-		return NULL;
-
-	pwnam.pw_name = pw_name;
-	pwnam.pw_gecos = pw_name;
-	pwnam.pw_dir = pw_dir;
-	pwnam.pw_gecos = pw_gecos;
-	pwnam.pw_shell = pw_shell;
-	pwnam.pw_passwd = pw_passwd;
-
-	while (fgets(buf, sizeof(buf), fp)) {
-		sscanf (buf, "%[^:]:%[^:]:%d:%d:%[^:]:%[^:]:%s\n",
-			pwnam.pw_name, pwnam.pw_passwd,
-			&pwnam.pw_uid, &pwnam.pw_gid,
-			pwnam.pw_gecos, pwnam.pw_dir, pwnam.pw_shell);
-		if (strcmp(pwnam.pw_name, name) == 0) {
-			fclose(fp);
-			return &pwnam;
+	int i, err = 0;
+	*uid = strtoul(intstr, NULL, 10);
+
+	if (readpwentry != 0) {
+		for (i = 0; intstr[i] != '\0'; ++i) {
+			if (!isdigit(intstr[i])) {
+				err = 1;
+				break;
+			}
 		}
 	}
-	fclose(fp);
-	return NULL;
+
+	return err;
```

Comment:
I think this condition `readentry != 0` is incorrect.
What will the `readuid()` return for such a record:
`name:passwd:0<EOF>`

`readpwentry = 1 // EOF`
`err = 0 // only digits`
and `readuid()` return `0` but record is invalid.

If you want the `readuid()` function to check if the field contains only digits then you can do it e.g. like this:
```
char *end;
if (!readpwentry) {
	*uid = strtoul(intstr, &end, 10);
	readpwentry = *end != '\0' || !isdigit(intstr[0]);
}
return readpwentry;
```

I think a better solution if the field contains only digits would be to check it in `getpwby()` function without an additional function likie this:
```
char *end;
if (readpwentry(fp, intstr, sizeof(intstr), 0)) {
	break;
}
else {
	*uid = strtoul(intstr, &end, 10);
	if (*end != '\0' || !isdigit(intstr[0])) {
		break;
	}
}
```
`isdigit()` at the end can be omitted .The `strtoul()` implementation in `libphoenix` only allows digits at `base = 10`.

Reply:
Yes, you are right. The part that includes `readpwenrty` was lost at last commit and scenario you have show was possible. The simple change is initialize `err` in `readuid` not with 0 but with `readpwentry` integer.
Thanks!